crates/p2p/src/client/peer_agnostic/parse.rs

```diff
@@ -646,3 +654,69 @@ pub(crate) mod events {
         impl_take_parsed_and_should_stop!(events);
     }
 }
+
+/// Parses (header, signature) pairs. Expects chunks(2) as input,
+/// will panic if given a chunk with length not one or two.
+///
+/// Errors if the chunk is not a (header, signature) pair or a
+/// successful [BlockHeadersResponsePart::Fin].
+pub(crate) fn handle_signed_header_chunk(
+    chunk: Vec<BlockHeadersResponsePart>,
+) -> Option<anyhow::Result<SignedBlockHeader>> {
+    use anyhow::anyhow;
+
+    if let [single] = chunk.as_slice() {
+        match single {
+            Header(_) => {
+                return Some(Err(anyhow!("Stream finalized with header")));
+            }
+            Signatures(_) => {
+                return Some(Err(anyhow!("Stream finalized with signature")));
+            }
+            Fin(p2p_proto::common::Fin { error: Some(error) }) => {
```

Comment:
`Fin::too_much` is not really an error if it follows at least one valid previous chunk; it indicates that the responding peer imposes an internal limit wrt query range

Reply:
I guess the same logic could be applied at least to this error:
- `Fin::unknown` - peer has a part of the queried range with some later blocks missing.

Reply:
ok, iiuc the outer loop handles this correctly: correct data up to some point is retained, the next peer is queried for the remaining part of the range